distsql: introduce fallbackRowContainer

asubiotto opened the pull request:
This change introduces a row container that stores rows in memory but falls back to disk if a memory limit is hit.

Closes #26107

Commit:
distsql: remove unused ctx parameter from makeDiskRowContainer

Release note: None

Commit:
distsql: add diskBackedRowContainer

This commit introduces a row container that stores rows in memory but
spills to disk if a memory limit is hit.

Closes #26107

Release note: None

Commit:
distsql: add diskBackedRowContainer to sortAllProcessor

This replaces the logic in the sortAllProcessor with a row container
that falls back to disk automatically.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26357)
<!-- Reviewable:end -->

asubiotto commented:
Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[pkg/sql/distsqlrun/row_container.go, line 241 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9jZp1dCKmY8zkrJc:-LEA9jZp1dCKmY8zkrJd:b-lu4wi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L241)):*
> ```Go
> // store rows and falls back to disk automatically if memory usage exceeds a
> // given budget.
> type fallbackRowContainer struct {
> ```

I don't love the name, any suggestions?

---

*[pkg/sql/distsqlrun/row_container.go, line 244 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9r7rCG_9fNDWjCKv:-LEA9r7rCG_9fNDWjCKw:b-nuf40u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L244)):*
> ```Go
> 	// src is the current sortableRowContainer that is being used to store rows.
> 	// All the sortableRowContainer methods are redefined rather than delegating
> 	// to an embedded struct because of how defer works:
> ```

Is this comment necessary?

---

*[pkg/sql/distsqlrun/row_container.go, line 334 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9tt88ECiIrDsEvSp:-LEA9tt88ECiIrDsEvSq:bvx7eu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L334)):*
> ```Go
> // UsingDisk returns whether or not the fallbackRowContainer is currently using
> // disk.
> func (f *fallbackRowContainer) UsingDisk() bool {
> ```

Not sure if this is the right name for the method. I can't decide on the tense of use because the caller might call this during or after execution, and both are valid.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26357)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Great work! I have a bunch of comments but overall this approach is very solid.

---

Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/row_container.go, line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAKfj70qzwCNbmMeNm:-LEAKfj70qzwCNbmMeNn:bg1oo5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5264c9992e2b68b379d7c9eb0f2eec14c41d64fc/pkg/sql/distsqlrun/row_container.go#L257)):*
> ```Go
> 	// fallbackEnabled specifies whether the fallbackRowContainer should fall back
> 	// to disk when encountering a memory limit.
> 	fallbackEnabled bool
> ```

When would this be false? If this is a plan-time parameter, shouldn't we just create a normal row container instead?

---

*[pkg/sql/distsqlrun/row_container.go, line 301 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAKwf47gTWOCwVXU0t:-LEAKwf47gTWOCwVXU0u:b-ejjeds) ([raw file](https://github.com/cockroachdb/cockroach/blob/5264c9992e2b68b379d7c9eb0f2eec14c41d64fc/pkg/sql/distsqlrun/row_container.go#L301)):*
> ```Go
> 			return err
> 		}
> 		log.VEventf(ctx, 2, "fell back to disk")
> ```

It'd be nice to have the requested amount of memory, current amount of memory and limit in this message.

---

*[pkg/sql/distsqlrun/row_container.go, line 241 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9jZp1dCKmY8zkrJc:-LEAKH5j-LHFteNQ6a8R:baou681) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L241)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't love the name, any suggestions?
</blockquote></details>

How about `diskBackedRowContainer`?

---

*[pkg/sql/distsqlrun/row_container.go, line 244 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9r7rCG_9fNDWjCKv:-LEAKLP75FkfGcLH5zcN:brlr5q5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L244)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is this comment necessary?
</blockquote></details>

Might as well leave it to prevent people from having to re-learn your lesson.

---

*[pkg/sql/distsqlrun/row_container.go, line 334 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9tt88ECiIrDsEvSp:-LEALCUg5kPXjIJSkwTA:ban8hj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L334)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Not sure if this is the right name for the method. I can't decide on the tense of use because the caller might call this during or after execution, and both are valid.
</blockquote></details>

`Spilled()`?

Also, it seems weird to do this with type introspection. I think I would be happier if the struct had three pointers: a memRowContainer, a diskRowContainer, and the current row container for use in AddRow and NewIterator. This feels more static to me. I think debugging the current structure might be difficult, because you have to use type introspection to figure out what the state of the system is.

Also, this opens the door to a potential future optimization of using both a disk and mem rowContainer at once. If batching writes to the disk container is desirable, then it would probably nice to fill up the memRowContainer for new rows even after we start to spill, so that we can write to disk in large batches.

---

*[pkg/sql/distsqlrun/sorter.go, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAM_1i2P7i80_CatV_:-LEAM_1i2P7i80_CatVa:b-fg1vjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/sorter.go#L174)):*
> ```Go
> 	// than the COCKROACH_WORK_MEM was reached. We should distinguish between
> 	// these cases and log the event to facilitate debugging of queries that
> 	// may be slow for this reason.
> ```

What do you mean by this?

---

*[pkg/sql/distsqlrun/sorter.go, line 183 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAMbHy1S5FiQB4nK5T:-LEAMbHy1S5FiQB4nK5U:b-kcb83s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/sorter.go#L183)):*
> ```Go
> 		rowContainerMon,
> 	)
> 	if !useTempStorage {
> ```

To follow up on the previous comment, you can use this opportunity to decide which kind of container to init, rather than have a DisableFallback method. e.g.

`if useTempStorage {proc.rows = &fallbackRowContainer} else {proc.rows = &memRowContainer}`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-:-LEAMykV-VS6z4lkyxFV:b-wxc3qe)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
I've also been thinking that these row containers should probably live in `sqlbase`. If you agree, I'll make that a separate PR

---

Review status: 6 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/row_container.go, line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAKfj70qzwCNbmMeNm:-LEBDEySDsh6KmGZ81Ce:b-jgjnei) ([raw file](https://github.com/cockroachdb/cockroach/blob/5264c9992e2b68b379d7c9eb0f2eec14c41d64fc/pkg/sql/distsqlrun/row_container.go#L257)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

When would this be false? If this is a plan-time parameter, shouldn't we just create a normal row container instead?
</blockquote></details>

When for example an operator disables the usage of external storage through one of the cluster settings. You're right, changed.

---

*[pkg/sql/distsqlrun/row_container.go, line 301 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAKwf47gTWOCwVXU0t:-LEBEKvbA3V6WCgbqhjM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5264c9992e2b68b379d7c9eb0f2eec14c41d64fc/pkg/sql/distsqlrun/row_container.go#L301)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It'd be nice to have the requested amount of memory, current amount of memory and limit in this message.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/row_container.go, line 241 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9jZp1dCKmY8zkrJc:-LEBDByP-t8ftjZljhZT:bt4ox1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L241)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

How about `diskBackedRowContainer`?
</blockquote></details>

Done. Thanks for the suggestion.

---

*[pkg/sql/distsqlrun/row_container.go, line 334 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEA9tt88ECiIrDsEvSp:-LEFM5YhA3sqX0QAvAwH:b-a72041) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/row_container.go#L334)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`Spilled()`?

Also, it seems weird to do this with type introspection. I think I would be happier if the struct had three pointers: a memRowContainer, a diskRowContainer, and the current row container for use in AddRow and NewIterator. This feels more static to me. I think debugging the current structure might be difficult, because you have to use type introspection to figure out what the state of the system is.

Also, this opens the door to a potential future optimization of using both a disk and mem rowContainer at once. If batching writes to the disk container is desirable, then it would probably nice to fill up the memRowContainer for new rows even after we start to spill, so that we can write to disk in large batches.
</blockquote></details>

Done. Also changed this file to refer to "spilling" rather than "falling back", it sounds better to me.

---

*[pkg/sql/distsqlrun/sorter.go, line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAM_1i2P7i80_CatV_:-LEFPYN-7cUSbBVxusVy:b-bgnhha) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/sorter.go#L174)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What do you mean by this?
</blockquote></details>

It was an old TODO from when I originally added external storage to the sorter. Basically just saying that there should be a way for users to understand that there are other memory limits other than the 64MB per-processor limit that could trigger a spill to disk (e.g. multiple concurrent queries) to avoid surprises. I think this TODO is more confusing than helpful so removed it.

---

*[pkg/sql/distsqlrun/sorter.go, line 183 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEAMbHy1S5FiQB4nK5T:-LEFQ50CBq-LCDbB1T1p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3709e65806e0b52322c5501c492b32450356e64d/pkg/sql/distsqlrun/sorter.go#L183)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

To follow up on the previous comment, you can use this opportunity to decide which kind of container to init, rather than have a DisableFallback method. e.g.

`if useTempStorage {proc.rows = &fallbackRowContainer} else {proc.rows = &memRowContainer}`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-:-LEBC7-p1gJcHyBMJeCW:b-awbkei)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, I think a few more tests could be good.

With regards to the packages, even better, why don't you make a new `rowcontainer` package for them? That would help keep their API clean and their tests fast.

---

Review status: 6 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/distsqlrun/row_container_test.go, line 143 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEFZMVvDEl0fX_Y2BpE:-LEFZMVvDEl0fX_Y2BpF:bsrks3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/310567a955343bbce88da860353f6f63c55559c6/pkg/sql/distsqlrun/row_container_test.go#L143)):*
> ```Go
> 		return &frc
> 	}
> ```

Can you add a test to make sure that it doesn't spill if the error isn't an out of memory error? Obviously the code works there today, but it's an important regression case.

---

*[pkg/sql/distsqlrun/row_container_test.go, line 164 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEFYn7G7BNR3yNjBqcc:-LEFYn7G7BNR3yNjBqcd:b-nqewwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/310567a955343bbce88da860353f6f63c55559c6/pkg/sql/distsqlrun/row_container_test.go#L164)):*
> ```Go
> 			t.Fatal("unexpectedly using disk")
> 		}
> 		frc.spillToDisk(ctx)
> ```

How come this test uses the internals of the row container to do the spilling? Wouldn't it be better to verify that the spilling happens correctly and automatically?

Also, can you test that you can successfully retrieve the rows *before* the container spills to disK?

---

*[pkg/sql/distsqlrun/row_container_test.go, line 205 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEFYSkU4KvRK0rD4KgU:-LEFYSkU4KvRK0rD4KgV:bkiwh51) ([raw file](https://github.com/cockroachdb/cockroach/blob/310567a955343bbce88da860353f6f63c55559c6/pkg/sql/distsqlrun/row_container_test.go#L205)):*
> ```Go
> 		defer diskMonitor.Stop(ctx)
> 
> 		frc := newDiskBackedRowContainer()
> ```

nit: these frc variable names don't make sense anymore.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-:-LEFUeIBCrQrWXxuPdb_:b7kb2m5)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
TFTR. I'm not sure what else I should test. I think that what your comments bring up are already tested, but let me know if you disagree.

I'm not sure a new `rowcontainer` package would be better than having them in `sqlbase`, but the only reason I think this is because row containers use a lot of stuff from `sqlbase` (they're a utility on top of `sqlbase` rows that are ordered according to properties defined in `sqlbase`).

---

Review status: 1 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/row_container_test.go, line 143 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEFZMVvDEl0fX_Y2BpE:-LEG7JZ8317f2S2BRojC:blom1t1) ([raw file](https://github.com/cockroachdb/cockroach/blob/310567a955343bbce88da860353f6f63c55559c6/pkg/sql/distsqlrun/row_container_test.go#L143)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can you add a test to make sure that it doesn't spill if the error isn't an out of memory error? Obviously the code works there today, but it's an important regression case.
</blockquote></details>

I think that verifying that we didn't spill to disk after adding a couple of rows in the `NormalRun` test covers this case.

---

*[pkg/sql/distsqlrun/row_container_test.go, line 164 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-LEFYn7G7BNR3yNjBqcc:-LEG82sY2FjG_KcxpDUb:b-uza5ve) ([raw file](https://github.com/cockroachdb/cockroach/blob/310567a955343bbce88da860353f6f63c55559c6/pkg/sql/distsqlrun/row_container_test.go#L164)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

How come this test uses the internals of the row container to do the spilling? Wouldn't it be better to verify that the spilling happens correctly and automatically?

Also, can you test that you can successfully retrieve the rows *before* the container spills to disK?
</blockquote></details>

Because the alternative is to figure out how much space rows are taking and we test an automatic spill below in `AddRowOutOfMem`. Added the check to verify correct retrieval of rows before spilling.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-:-LEGAUo_21j7ANW4qpPT:b-a7b3ti)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

You'd still get benefits from having `rowcontainer` be a separate package even though it depends on stuff from `sqlbase`, i think. It's always nice to have small, well-contained packages, even if they have dependencies. It makes it clear what their interfaces are.

---

Review status: 1 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26357#-:-LELotN-4ajZZ4WZh6iR:b-m3eus2)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
I'll definitely give it a try!
bors r+
